README.md

```diff
@@ -110,37 +109,54 @@ Next run a suite of tests with `make test` and see a corresponding walkthrough
 of API features in [tests/demo/walkthrough.sh](tests/demo/walkthrough.sh). For
 more demonstrations, see the [demos](demos) directory.
 
-## Running the Registry API server locally
+## Running the Registry API server
+
+### Configuration
+
+Configuration for `registry-server` is loaded from the
+`$HOME/.config/registry/` directory. YAML, JSON, TOML, and other common formats
+are supported. The file name must be `registry-server` with a file extension
+indicating the configuration format, e.g. `registry-server.yaml`. Refer to the
+[example](config/registry-server.yaml) for documentation of each
+configurable value.
+
+Values can be set or overridden using environment variables. Each configuration
+value should be uppercased, prefixed with `REGISTRY_`, and use underscores to
+indicate nesting. For example, you can override the `database.driver` value by
+setting the `REGISTRY_DATABASE_DRIVER` environment variable.
 
 ### Running the Registry API server
 
-Running `source auth/LOCAL.sh` will configure your environment to run the
-Registry API server locally and for the included clients to call your local
-instance. Start the server by running `registry-server`. By default a SQLite
-backend will be used (this is equivalent to running
-`registry-server -c config/sqlite.yaml`).
+Run `source auth/LOCAL.sh` to configure your environment to run the Registry
+API server locally and for the included clients to call your local instance.
+Start the server by running `registry-server`.
 
-### Optional: Running with a PostgreSQL backend
+### Optional: Use a PostgreSQL database on the local machine
 
-The `config` directory contains examples of files that can be used to configure
-the `registry-server`. [config/postgres.yaml](config/postgres.yaml) contains a
-sample configuration; you'll likely need to customize this for your own
-`postgresql` instance. Use it with:
+Ensure you have PostgreSQL [installed](https://www.postgresql.org/download/)
+and set up on your machine. After it's ready, update the `database.driver` and
+`database.dsn` values in your configuration.
```

Comment:
>I don't think we get much value from using "dsn" as the configuration name here. I am aware of its origin, but I suspect many people will have to look it up.

I think that's a good thing because every possible reference to the config value gives users enough information to find out what exactly what the expected format is. The value is discoverability IMO.

>Could we instead use "config" and explain in the example YAML file that it's a "data source name"?

There is an explanation in the example config file because I think both "dsn" and "config" require explanation. 

Suppose someone is setting up the server's docker image and sees the following command. If the user wants to know the expected format for the host field, how likely are they to find their way to this information?

```
docker run \
  -p 8080:8080 \
  -e REGISTRY_DATABASE_DRIVER=postgres \
  -e REGISTRY_DATABASE_CONFIG="host=${PGHOST} port=5432 user=registry dbname=registry password=iloveapis sslmode=disable" \
  registry-server:latest
```

Using the name `REGISTRY_DATABASE_DSN` instead, is the user more or less likely to succeed in connecting to their database? I can only think of more ways to finding the correct information.

>Since ".dsn" is scoped to "database", "database.config" wouldn't conflict with anything else, and superficially, it looks better ("config" is the same length as "driver") and isn't a somewhat-obscure acronym. (it's also easier to pronounce and less likely to be mistyped - the way I've seen people type "gprc", for example)

I agree that neither name will conflict with anything, and that "config" is a name that's nicer, more aesthetic, less obscure, easier to pronounce, and harder to mistype. How do you weigh these benefits compared to the increased discoverability of a name that includes "dsn" in it?

Reply:
"DSN" is almost always going to require a search, and a search for "REGISTRY_DATABASE_CONFIG" will more specifically lead to our docs (as would "REGISTRY_DATABASE_DSN").

There are two different questions that could require search. The first is "what is DSN?" (it's a configuration string, which is clear if we call it "config") The second is "what is the format of the config string?" Lookups are going to be needed for that regardless.

If we used "config",  could someone use this without knowing the definition of DSN? - I think yes, by following our examples. 

if we used "dsn", could someone use this without knowing the definition of DSN? - I think not, because they wouldn't even know it was a configuration string until they looked up DSN.

Reply:
>"DSN" is almost always going to require a search, and a search for "REGISTRY_DATABASE_CONFIG" will more specifically lead to our docs (as would "REGISTRY_DATABASE_DSN").

We might agree here. Can you clarify and provide an example if you're saying these are meaningfully different? Particularly the part about searching.

>There are two different questions that could require search. The first is "what is DSN?" (it's a configuration string, which is clear if we call it "config") The second is "what is the format of the config string?" Lookups are going to be needed for that regardless.

I'm not seeing your point here about the configuration strings. Can you provide an example of a scenario where this will matter? I agree that searching will be necessary to find the format of the string regardless of name.

>If we used "config", could someone use this without knowing the definition of DSN? - I think yes, by following our examples.

Agreed. Although I would say the same if we use any other name.

>if we used "dsn", could someone use this without knowing the definition of DSN? - I think not, because they wouldn't even know it was a configuration string until they looked up "DSN". So I think this is more burdensome for the typical user.

I think yes, by following our examples. Can you describe a situation where somebody will be able to find examples and follow them with the _CONFIG suffix but won't be able to with the _DSN suffix? I'm having a hard time imagining this scenario.

Does your opinion change with "connection string" in the name instead of "dsn"? I'm not sure if a DSN that isn't also a database connection string would actually work if provided.

Reply:
So I guess I think that these are overriding benefits: 

""config" is a name that's nicer, more aesthetic, less obscure, easier to pronounce, and harder to mistype."

I think of "dsn" as a type, like "float" or "string", and "config" is the role that it has in our system. But ultimately I'm prioritizing nice-looking configuration that doesn't surprise people with things that make them want to stop and look something else up on their first read through our docs... and names that are memorable, so that people don't have failures due to accidentally setting environment variables named "REGISTRY_DATABASE_DBN", "REGISTRY_DATABASE_DNS", or some other easily-missed misspelling.

Reply:
I think this is ultimately about the appropriateness of an acronym. To me "dsn" has nearly all the problems that [Rebekah Barry describes here](https://blog.prototypr.io/reasons-to-avoid-using-acronyms-e1a97c8ea6c4). The other aspects we've discussed are not irrelevant, but are much less important.

There's lots of other discussion of this online, but I think [Patrick Robert Doyle's thoughts](https://uxdesign.cc/just-fyi-acronyms-are-hurting-your-ux-and-roi-423572d69195) are worth noting, and I think this is relevant to documentation, UX, and [code](https://github.com/apigee/registry/pull/191#issuecomment-863746052).

Reply:
Agreed. I just don't think the problems associated with acronyms (and this specific acronym) outweigh the benefits associated with using standard and precise terminology. 

Reading Rebekah Barry's article, I get the impression that full expansions of acronyms (e.g. data source name) are preferred even at the cost of verbosity. I'm struggling to make the leap from avoiding acronyms to replacing them with terms like "config" that only vaguely describe the subject.

Reading Patrick Doyle's article, I get a much stronger affirmation of the same point.

>ask yourself “What does my audience call the thing I’m talking about?”

>Does a term already exist for the thing I am describing? Think about how you’ve been referring to it with others already.

>Try to avoid using acronyms and substitute them for either their full definition, or if you’re in a position to fix it, find a shorter and more descriptive term.

We're talking about data source names, so our audience either calls it a DSN, data source name, or they likely don't have any concept of it. The libraries we depend on refer to it as DSN. ODBC documentation going back to the 1980s refer to it as DSN or data source name. We can't fix those things, and "config" certainly isn't more descriptive.

--

Now I'm convinced to use "data source name" instead of "DSN," and I'm curious what you think of that since your concerns seem focused on issues with "DSN" that are mostly unique to it being an acronym. If you insist on "config" I can push up that change next week, but I think I've clearly stated why I think it's a poor choice.

Reply:
Thanks for reading and thinking about this. I think we should try to get to a consensus. Here are the alternatives side-by-side:

1. REGISTRY_DATABASE_DSN
2. REGISTRY_DATABASE_CONFIG
3. REGISTRY_DATABASE_DATA_SOURCE_NAME

We should consider if precision is worth the pedantry of asking people to use "REGISTRY_DATABASE_DATA_SOURCE_NAME" instead of "REGISTRY_DATABASE_CONFIG".

Also from Patrick Robert Doyle: "You’re guilty of trying to look smart. Look. Acronyms and initialisms make you look smart and informed. I’m guilty of that, you’re guilty of that." 

Every person who approaches our configuration is going to have a general idea of what "REGISTRY_DATABASE_CONFIG" does. Some will be puzzled by "REGISTRY_DATABASE_DATA_SOURCE_NAME", and some of these people will be distracted by the fact that we taught them a new term. Will they appreciate that? Is teaching database terminology our job?

Reply:
Thanks for reading and thinking about this. I think we should try to get to a consensus. Here are the alternatives side-by-side:

1. REGISTRY_DATABASE_DSN
2. REGISTRY_DATABASE_CONFIG
3. REGISTRY_DATABASE_DATA_SOURCE_NAME

We should consider if precision is worth the pedantry of asking people to use "REGISTRY_DATABASE_DATA_SOURCE_NAME" instead of "REGISTRY_DATABASE_CONFIG".

Also from Patrick Robert Doyle: "You’re guilty of trying to look smart. Look. Acronyms and initialisms make you look smart and informed. I’m guilty of that, you’re guilty of that." 

Every person who approaches our configuration is going to have a general idea of what "REGISTRY_DATABASE_CONFIG" does. Some will be puzzled by "REGISTRY_DATABASE_DATA_SOURCE_NAME", and some of these people will be distracted by the fact that we taught them a new term. Will they appreciate that? Is teaching database terminology our job?